geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/PingOpDistributedTest.java

```diff
@@ -0,0 +1,219 @@
+
+    client = getVM(0);
+    server1 = getVM(1);
+    server2 = getVM(2);
+    server1Port = ports[0];
+    server2Port = ports[1];
+    server1.invoke(() -> initServer(server1Port));
+    server2.invoke(() -> initServer(server2Port));
+
+    clientLogFile = folder.newFile("client-archive.log");
+
+  }
+
+  void parametrizedSetUp(String poolName, List<Integer> serverPorts) {
+    client.invoke(() -> initClient(poolName, serverPorts));
+  }
+
+  public void executePing(String poolName, int serverPort,
+      InternalDistributedMember distributedMember) {
+    PoolImpl poolImpl = (PoolImpl) PoolManager.find(poolName);
+    PingOp.execute(poolImpl, new ServerLocation("localhost", serverPort), distributedMember);
+  }
+
+  public Long getSingleHeartBeat() {
+    ClientHealthMonitor chm = ClientHealthMonitor.getInstance();
+    assertThat(chm.getClientHeartbeats()).isNotEmpty().hasSize(1);
+
+    return chm.getClientHeartbeats().entrySet().iterator().next().getValue();
+  }
+
+  @Test
+  public void regularPingFlow() {
+    final String poolName = testName.getMethodName();
+    parametrizedSetUp(poolName, Collections.singletonList(server1Port));
+    InternalDistributedMember distributedMember1 = (InternalDistributedMember) server1
+        .invoke(() -> cacheRule.getCache().getDistributedSystem().getDistributedMember());
+
+    server1.invoke(() -> {
+      ClientHealthMonitor chm = ClientHealthMonitor.getInstance();
+      assertThat(chm.getClientHeartbeats()).isEmpty();
+    });
+
+    client.invoke(() -> executePing(poolName, server1Port, distributedMember1));
+    Long firstHeartbeat = server1.invoke(this::getSingleHeartBeat);
+
+    client.invoke(() -> executePing(poolName, server1Port, distributedMember1));
+    Long secondHeartbeat = server1.invoke(this::getSingleHeartBeat);
+
+    assertThat(secondHeartbeat).isGreaterThan(firstHeartbeat);
+  }
+
+  @Test
+  public void memberShouldCorrectlyRedirectPingMessage() throws IOException {
+    final String poolName = testName.getMethodName();
+    parametrizedSetUp(poolName, asList(server1Port, server2Port));
+    InternalDistributedMember distributedMember1 = (InternalDistributedMember) server1
+        .invoke(() -> cacheRule.getCache().getDistributedSystem().getDistributedMember());
+    InternalDistributedMember distributedMember2 = (InternalDistributedMember) server2
+        .invoke(() -> cacheRule.getCache().getDistributedSystem().getDistributedMember());
+
+    asList(server1, server2).forEach(vm -> vm.invoke(() -> {
+      ClientHealthMonitor chm = ClientHealthMonitor.getInstance();
+      assertThat(chm.getClientHeartbeats()).isEmpty();
+    }));
+
+    System.out.println("########## FIRST PING START");
+    client.invoke(() -> executePing(poolName, server1Port, distributedMember2));
+    Long firstHeartbeatServer1 = server1.invoke(this::getSingleHeartBeat);
+    Long firstHeartbeatServer2 = server2.invoke(this::getSingleHeartBeat);
```

Comment:
@jujoramos There is something I dont get here: if there is a ClientHealthMonitor per server, how it is possible that a ping send for distributedMember2 creates an entry in the ClientHealthMonitor of server1? I have been debugging this and I only see one call in ClientHealthMonitor::receivedPing in server 2 (as expected) after executing the ping (line 164). Could it be a problem of the testing framework? Or am I missing something?

Reply:
Hey @alb3rtobr 
Sorry for the confusion here, the test `memberShouldCorrectlyRedirectPingMessage` is not finished at all, I was just using it to debug the issue.

Reply:
Ok! Anyway I found a possible cause for the unexpected ping. I think the assert in line 160 is not correct. Sometimes the test fails there, although no ping has been sent yet. The `ServerConnection::processHandshake` method calls `ClientHealthMonitor::registerClient`, which will cause that `ClientHealthMonitor.getClientHeartbeats()` will not be empty. I think sometimes the test is fast enough to pass the assert, and the heartbeat got in line 165 is the one created by `ServerConnection::processHandshake`

Reply:
Yeah, do not worry about this particular test method, the one showing the issue is `memberShouldNotRedirectPingMessageWhenClientCachedViewIdIsWrong`. You can use the other two as starting points to develop your own tests for the `GEODE-7565`.